<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Async/Await</title>
  </head>
  <body>
    <h1>Funciones Asíncronas (Async / Await)</h1>
    <script>
      function cuadradoPromise(value) {
        if (typeof value !== "number")
          return Promise.reject(
            `Error, el valor "${value}" ingresado no es un número`
          );
        return new Promise((resolve, reject) => {
          setTimeout(() => {
            resolve({ value, result: Math.pow(value, 2) });
          }, 0 | (Math.random() * 1000));
        });
      }
      async function funcionAsincronaDeclarada() {
        try {
          console.log("Inicio Async Function");
          let objeto = await cuadradoPromise(0);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(1);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(2);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(3);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(4);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(5);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          console.log("Fin Async Function");
        } catch (error) {
          console.error(error);
        }
      }
      funcionAsincronaDeclarada();

      const funcionAsincronaExpresada = async () => {
        try {
          console.log("Inicio Async Function");
          let objeto = await cuadradoPromise(6);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(7);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(8);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(9);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(10);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          objeto = await cuadradoPromise(11);
          console.log(`Async function: ${objeto.value}, ${objeto.result}`);
          console.log("Fin Async Function");
        } catch (error) {
          console.error(error);
        }
      };
      funcionAsincronaExpresada();
    </script>
  </body>
</html>